stdlib/list.ml

```diff
@@ -549,6 +549,29 @@ let rec compare_length_with l n =
       compare_length_with l (n-1)
 ;;
 
+(** {1 Comparison} *)
+
+(* Note: we are *not* shortcutting the list by using
+   [List.compare_lengths] first; this may be slower on long lists
+   immediately start with distinct elements. It is also incorrect for
+   [compare] below, and it is better (principle of least surprise) to
+   use the same approach for both functions. *)
+let rec equal eq l1 l2 =
+  match l1, l2 with
+  | [], [] -> true
+  | [], _::_ | _::_, [] -> false
+  | a1::l1, a2::l2 -> eq a1 a2 && equal eq l1 l2
+
+let rec compare cmp l1 l2 =
+  match l1, l2 with
+  | [], [] -> 0
+  | [], _::_ -> -1
+  | _::_, [] -> 1
+  | a1::l1, a2::l2 ->
+    let c = cmp a1 a2 in
+    if c <> 0 then c
+    else compare cmp l1 l2
+
 (** {1 Iterators} *)
```

Comment:
@gasche I've just seen the addition of these two functions (on discuss). Depending of the usage contexts and how lists are constructed, I was wondering if it's not a good idea to add pre-comparison l1 == l2 to immediately answer when the two lists are physically equal.

Reply:
Agree. Maybe the code I interact with is odd, but the profiler has taught me to always start structural comparison functions like these with a physical equality test. If it doesn't help the cost is usually nothing and at worst very small. But on the other hand the savings on inputs with some internal sharing can be very significant.

Reply:
This optimization would only be valid if the `cmp` parameter is reflexive (it always returns `true` on equal inputs), for example `List.compare (fun _ _ -> false) [()] [()]` changes behavior. This makes me inclined to think that we may not want this trick. (As far as I know, the polymorphic-equality usage in the standard library are always valid, they do not require extra assumptions.)

Reply:
Ah, ok, fair point. I'm implicitly assuming that the compare functions should be at least preorders. I'm not familiar with / accustomed to uses where they are not reflexive.

Reply:
Ok, I understand.

Reply:
One can use `List.equal` as a variant of `List.for_all2` that behaves nicely on different-length lists, and then the predicate may be arbitrary. I don't really like this usage either, but I don't have a much better proposal to offer for those use-cases.

Reply:
> This optimization would only be valid if the `cmp` parameter is reflexive (it always returns `true` on equal inputs), for example `List.compare (fun _ _ -> false) [()] [()]` changes behavior. This makes me inclined to think that we may not want this trick. (As far as I know, the polymorphic-equality usage in the standard library are always valid, they do not require extra assumptions.)

I can understand the argument for the `equal` function, but I don't think it applies as well to `compare`. Given that it is explicitly documented that `x == y` implies `Stdlib.compare x y = 0`, I would expect other `compare` functions from the standard library to provide similar guarantees.
(If I remember correctly, with custom values you can get cases where `Stdlib.compare x x` would have returned a non-zero result if the physical equality shortcut didn't trigger, so the fact that `List.compare` is taking a function as argument, and that argument is not guaranteed to be a total order, is not necessarily a fundamental difference with `Stdlib.compare`).
On the other hand, this does change the semantics for comparison on infinite lists. If I was asked what should be the semantics of `let rec l = 0 :: l in List.compare Stdlib.compare l l` I would be fine with both returning `0` and looping forever. If you prefer the looping version that would make a decent argument for keeping `compare` as it is.

Reply:
I agree that it makes more sense for `compare`. (The snippet I wrote used `compare` but it was a mistake, I rewrote it with `equal` as originally intended.) Maybe you want to submit a PR with this argument?

Reply:
Done: https://github.com/ocaml/ocaml/pull/10239